kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala

```diff
@@ -230,4 +232,16 @@ private[kyuubi] class EngineRef(
         create(discoveryClient, extraEngineLog)
       }
   }
+
+  def close(): Unit = {
+    if (shareLevel == CONNECTION && builder != null) {
+      try {
+        builder.shutdown()
+        engineManager.killApplication(builder.clusterManager(), engineRefId)
```

Comment:
the ordering is a little confused: we have closed the builder, but get `clusterManager` from it again. It seems better:
```scala
val clusterManager = builder.clusterManager()
builder.close()
engineManager.killApplication(clusterManager)
```